### PR TITLE
Add QThread, QLock and another Qt sync object

### PR DESCRIPTION
**Pull request by nk0leg**
Add QThread, QLock and another Qt sync object

see SDLWIN-133

### CODE REVIEW

**Comment by nk0leg** [comment]
@pestOO , @malirod , @anosach-luxoft  please review (duplicate #74)

**Comment by pestOO** [comment]
@nk0leg, It's not possible for me today to review this PR.
If I'm blocking some other activities - don't wait me.
If this review will not be closed at Monday evening - I'll review it.

**Comment by AGaliuzov** [comment]
reviwed

**Comment by nk0leg** [comment]
@malirod , @anosach-luxoft  pull request has been updated. Please look 627e1ea

File: src/components/include/utils/lock.h [hunk 1]
```diff
@@ -62,16 +67,26 @@ class SpinMutex {
...
       return;
     }
     for (;;) {
 #if defined(OS_POSIX)
       sched_yield();
-#elif defined(OS_WINDOWS)
+#elif defined(WIN_NATIVE)
       SwitchToThread();
+#elif defined(QT_PORT)
+    QThread::yieldCurrentThread();
```
Comment (malirod) [inline src/components/include/utils/lock.h:83]: seems there is wrong indent

File: src/components/include/utils/threads/thread_delegate.h [hunk 2]
```diff
@@ -53,6 +61,10 @@ class ThreadDelegate {
    */
   virtual void threadMain() = 0;
 
+#ifdef QT_PORT
+  Q_SIGNAL void exit_thread();
```
Comment (malirod) [inline src/components/include/utils/threads/thread_delegate.h]: As far as I understand, this signal is our custom signal. Why you don't use our code style for the name? Or I miss something?

File: src/components/include/utils/lock.h (hunk 1 above)
Comment (nk0leg) [inline src/components/include/utils/lock.h:83]: see a3679a0

File: src/components/include/utils/threads/thread.h [hunk 3]
```diff
@@ -91,7 +96,14 @@ void sleep(uint32_t ms);
 Thread* CreateThread(const char* name, ThreadDelegate* delegate);
 void DeleteThread(Thread* thread);
 
+#if defined(QT_PORT)
+class Thread : public QObject {
+  Q_OBJECT
+ public slots:
+  void thread_cancelled_exit();
```
Comment (malirod) [inline src/components/include/utils/threads/thread.h]: this name seems to be invalid too

**Comment by anosach-luxoft** [comment]
Reviewed

**Comment by malirod** [comment]
@pestOO, please review once more.

**Comment by nk0leg** [comment]
@pestOO , @malirod , @anosach-luxoft please look 45d6fac

File: src/components/include/utils/threads/thread_delegate.h [hunk 4]
```diff
@@ -53,6 +61,10 @@ class ThreadDelegate {
    */
   virtual void threadMain() = 0;
 
+#ifdef QT_PORT
+  Q_SIGNAL void ExitThread();
```
Comment (anosach-luxoft) [inline src/components/include/utils/threads/thread_delegate.h]: @nk0leg remove this line

**Comment by anosach-luxoft** [comment]
Reviewed

File: src/components/include/utils/threads/thread_delegate.h (hunk 4 above)
Comment (nk0leg) [inline src/components/include/utils/threads/thread_delegate.h]: see 45d6fac

**Comment by anosach-luxoft** [comment]
@nk0leg I have an issue with this implementation. Steps are follows:
- Run SDL
- Connect mobile (any transport)
- Register any application and put it into FULL level

After 10 - 15 seconds you'll get ASSERT `Locking already taken not recursive mutex`. Seems like we try to lock non-recursive mutex twice

**Comment by nk0leg** [comment]
Your sequencing is not reproduced.
It is necessary to use cleared build.

**Comment by anosach-luxoft** [comment]
Did you try `WIN_NATIVE` build?

**Comment by nk0leg** [comment]
Not, I didn't.
You try?

**Comment by anosach-luxoft** [comment]
I've reproduced issue described above on `WIN_NATIVE` build

**Comment by nk0leg** [comment]
see 688b26d

**Pull request merged**